BlobSidecarsByRoot

kasey opened the pull request:
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

This implements the `BlobSidecarsByRoot` p2p RPC handler according to [deneb/p2p-interface.md](https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/p2p-interface.md#blobsidecarsbyroot-v1).

terencechain commented on config/fieldparams/mainnet.go:
```diff
@@ -26,4 +26,6 @@ const (
 	SyncAggregateSyncCommitteeBytesLength = 64            // SyncAggregateSyncCommitteeBytesLength defines the length of sync committee bytes in a sync aggregate.
 	MaxWithdrawalsPerPayload              = 16            // MaxWithdrawalsPerPayloadLength defines the maximum number of withdrawals that can be included in a payload.
+	BlobSize                              = 131072        // defined to match blob.size in bazel ssz codegen
+	MaxBlobsPerBlock                      = 4             // MAX_BLOBS_PER_BLOCK
```
I originally had `MaxBlobsPerBlock` under beacon config. Looking at this, defining in fieldparam is a better place than config. It is a preset and preset is closer to fieldparams than beacon config

terencechain commented on config/params/mainnet_config.go:
```diff
@@ -284,14 +286,17 @@ func FillTestVersions(c *BeaconChainConfig, b byte) {
 	c.AltairForkVersion = make([]byte, fieldparams.VersionLength)
 	c.BellatrixForkVersion = make([]byte, fieldparams.VersionLength)
 	c.CapellaForkVersion = make([]byte, fieldparams.VersionLength)
+	c.DenebForkVersion = make([]byte, fieldparams.VersionLength)
```
Thanks for taking care all the configs!

kasey replied:
TODO: I think if a user-supplied yaml config doesn't specify a deneb fork version, we could fail to load it, because the zero-value for the version could conflict. I need to test this.

prestonvanloon commented on beacon-chain/sync/rpc_blob_sidecars_by_root.go:
```diff
@@ -0,0 +1,86 @@
+		return errors.New("message is not type BlobSidecarsByRootReq")
+	}
+	minReqEpoch := blobMinReqEpoch(s.cfg.chain.FinalizedCheckpt().Epoch, slots.ToEpoch(s.cfg.clock.CurrentSlot()))
+	blobIdents := *ref
+	for i := range blobIdents {
```
Before you start work processing this list, should you validate that it is within the max bounds?

> No more than MAX_REQUEST_BLOBS_SIDECARS * MAX_BLOBS_PER_BLOCK may be requested at a time.

kasey replied:
definitely, thanks for pointing that out, I'm surprised there is no check at all for this, good catch.

kasey replied:
updated -- wound up sorting request elements to avoid `call the DB lookup multiple times for each index for a given root.`

prestonvanloon commented on beacon-chain/sync/rpc_blob_sidecars_by_root.go:
```diff
@@ -0,0 +1,86 @@
+		return errors.New("message is not type BlobSidecarsByRootReq")
+	}
+	minReqEpoch := blobMinReqEpoch(s.cfg.chain.FinalizedCheckpt().Epoch, slots.ToEpoch(s.cfg.clock.CurrentSlot()))
+	blobIdents := *ref
+	for i := range blobIdents {
```
It might also be a good idea to add a ctx expired check at the start of the loop routine.

```suggestion
	blobIdents := *ref
	for i := range blobIdents {
		if err := ctx.Err(); err != nil {
			return err
		}
```

kasey replied:
done

prestonvanloon commented on beacon-chain/sync/rpc_blob_sidecars_by_root.go:
```diff
@@ -0,0 +1,86 @@
+		}
+		s.rateLimiter.add(stream, 1)
+	}
+	closeStream(stream, log)
```
Shouldn't this be a defer as soon as the stream is opened? Seems like any of the early terminations on error wouldnt close the stream. 

Also, wouldn't the caller be responsible for closing the stream? Perhaps you could add some commentary or something to explain why this is here in the way that it is.

kasey replied:
If you look at other RPC handlers, they all do the same thing. The reason is that the error handling functions independently call `closeStream`, so calling it in a defer would cause the stream to be closed twice. I want to refactor the dispatch code to take care of this after the handler completes and remove the close from the error handlers, since we should always close the stream in both error and successful paths.

kasey replied:
I'm going to do some work to fix this for all RPCs in develop. In the meantime the context check adds another place where we need to explicitly call closeStream. I thought this was cleaner than my other idea, which was to defer with an enclosed bool to track whether an error method had been called. Once I get the other PR into develop and we rebase, I'll update this code to defer the stream close. Hopefully that doesn't block this PR.

prestonvanloon commented on beacon-chain/sync/rpc_blob_sidecars_by_root.go:
```diff
@@ -0,0 +1,86 @@
+			return types.ErrBlobLTMinRequest
+		}
+		SetStreamWriteDeadline(stream, defaultWriteDuration)
+		if chunkErr := WriteBlobSidecarChunk(stream, s.cfg.chain, s.cfg.p2p.Encoding(), sc); chunkErr != nil {
+			log.WithError(chunkErr).Debug("Could not send a chunked response")
+			s.writeErrorResponseToStream(responseCodeServerError, types.ErrGeneric.Error(), stream)
+			tracing.AnnotateError(span, chunkErr)
+			return chunkErr
+		}
+		s.rateLimiter.add(stream, 1)
```
Want to increment this before you write the chunk so that it is incremented whether or not the response was successful? 

Suggestion below moves `s.rateLimiter.add(stream, 1)` to the line before the if statement.

```suggestion
		s.rateLimiter.add(stream, 1)
		if chunkErr := WriteBlobSidecarChunk(stream, s.cfg.chain, s.cfg.p2p.Encoding(), sc); chunkErr != nil {
			log.WithError(chunkErr).Debug("Could not send a chunked response")
			s.writeErrorResponseToStream(responseCodeServerError, types.ErrGeneric.Error(), stream)
			tracing.AnnotateError(span, chunkErr)
			return chunkErr
		}
```

kasey replied:
yeah that makes sense, actually I will move this to right after the DB lookup.

nisdas replied:
Yeap, was about to also add that suggestion too ^

kasey replied:
done

nisdas commented on beacon-chain/sync/rate_limiter.go:
```diff
@@ -59,6 +59,9 @@ func newRateLimiter(p2pProvider p2p.P2P) *limiter {
 	blockCollectorV2 := leakybucket.NewCollector(allowedBlocksPerSecond, allowedBlocksBurst, blockBucketPeriod, false /* deleteEmptyBuckets */)
 
+	// for BlobSidecarsByRoot and BlobSidecarsByRange
+	blobCollector := leakybucket.NewCollector(allowedBlocksPerSecond, allowedBlocksBurst, blockBucketPeriod, false)
```
is it ok to assume 1-1 mapping for blob rate limiter params( same as for a block).

kasey replied:
Good point, it's probably better for this to have its own set of params that can be tuned independently.

nisdas commented on beacon-chain/sync/rpc_send_request.go:
```diff
@@ -129,3 +134,65 @@ func SendBeaconBlocksByRootRequest(
+		if !found {
+			return nil, errors.Wrapf(ErrBlobUnmarshal, fmt.Sprintf("unrecognized fork digest %#x", ctxb))
+		}
+		if v != version.Deneb {
```
this would break for any forks after deneb

kasey replied:
This was intentional, because this function returns a deneb-specific type. I thought it was premature to add an interface type for blobs, so it seemed natural to make this break in an obvious way for future forks, so that we'll know we need to update this code to handle multiple forks.

nisdas commented on config/params/network_config.go:
```diff
@@ -17,6 +17,7 @@ type NetworkConfig struct {
 	BlobsidecarSubnetCount           uint64           `yaml:"BLOBSIDECAR_SUBNET_COUNT"`              // BlobsidecarSubnetCount is the number of blobsidecar subnets used in the gossipsub protocol.
 	AttestationPropagationSlotRange  primitives.Slot  `yaml:"ATTESTATION_PROPAGATION_SLOT_RANGE"`    // AttestationPropagationSlotRange is the maximum number of slots during which an attestation can be propagated.
 	MaxRequestBlocks                 uint64           `yaml:"MAX_REQUEST_BLOCKS"`                    // MaxRequestBlocks is the maximum number of blocks in a single request.
+	MaxRequestBlocksDeneb            uint64           `yaml:"MAX_REQUEST_BLOCKS_DENEB"`              // MaxRequestBlocksDeneb is the maximum number of blocks in a single request after the Deneb fork.
```
why do we need a new param after deneb

nisdas commented on beacon-chain/sync/rpc_send_request.go:
```diff
@@ -129,3 +134,65 @@ func SendBeaconBlocksByRootRequest(
+		if err := decode(stream, sc); err != nil {
+			return nil, errors.Wrap(err, "failed to decode the protobuf-encoded BlobSidecar message from RPC chunk stream")
+		}
+		sidecars = append(sidecars, sc)
```
there seems to be no validation on the blob sidecars, how do we verify that the blobs being provided are the ones we want ?

kasey replied:
The response may not include all requested values:
```
Clients MUST respond with at least one sidecar, if they have it. Clients MAY limit the number of blocks and sidecars in the response.
```
Since the caller of this function may need to formulate a new request to get the missing values, I thought it was better to leave that kind of logic to the user of the function. What do you think?

nisdas replied:
I think that is fine, but is there any correctness check that we can do to verify that the sidecars provided by the peer are at least a subset of what we requested ? Thinking about how to protect us in the event of malicious peer just returning us nonsense

kasey commented on beacon-chain/p2p/rpc_topic_mappings.go:
```diff
@@ -113,6 +123,19 @@ var versionMapping = map[string]bool{
+// Phase0 did not have the notion of context bytes, which prefix wire-encoded values with a [4]byte identifier
+// to convey the schema for the receiver to use. These RPCs had a version bump to V2 when the context byte encoding
+// was introduced. For other RPC methods, context bytes are always required.
+var OmitContextBytesV1 = map[string]bool{
```
My thought was that not having context bytes is an exception to the rule, and any new endpoints will have context bytes. Doing it this way means we don't need to update the map to add a new endpoint.

kasey commented on beacon-chain/sync/blobs_test.go:
```diff
@@ -0,0 +1,336 @@
+	code, _, err := ReadStatusCode(stream, encoding)
+	require.NoError(t, err)
+	require.Equal(t, r.code, code, "unexpected response code")
+	// require.Equal(t, r.message, msg, "unexpected error message")
```
thanks!

kasey replied:
updated

kasey commented on beacon-chain/p2p/types/types.go:
```diff
@@ -120,3 +121,66 @@ func (m *ErrorMessage) UnmarshalSSZ(buf []byte) error {
+	bufLen := len(buf)
+	maxLength := int(params.BeaconNetworkConfig().MaxRequestBlobsSidecars) * blobIdSize
+	if bufLen > maxLength {
+		return errors.Errorf("expected buffer with length of upto %d but received length %d", maxLength, bufLen)
```
updated

prestonvanloon replied:
Looks un-updated, Could you please check?

kasey replied:
you are correct, fixed here https://github.com/prysmaticlabs/prysm/pull/12420/commits/0a62357c6ee3c3b934db11a130c2ca4e8b1026d0

terencechain commented on beacon-chain/p2p/types/types.go:
```diff
@@ -120,3 +124,90 @@ func (m *ErrorMessage) UnmarshalSSZ(buf []byte) error {
+	return append(dst, marshalledObj...), nil
+}
+
+// MarshalSSZ Marshals the block by roots request type into the serialized object.
```
```suggestion
// MarshalSSZ Marshals the blob sidecars by roots request type into the serialized object.
```

terencechain commented on beacon-chain/p2p/types/types.go:
```diff
@@ -120,3 +124,90 @@ func (m *ErrorMessage) UnmarshalSSZ(buf []byte) error {
+	return len(*b) * blobIdSize
+}
+
+// MarshalSSZTo marshals the block by roots request with the provided byte slice.
```
```suggestion
// MarshalSSZTo marshals the blob sidecars by roots request with the provided byte slice.
```

terencechain commented on beacon-chain/p2p/types/types.go:
```diff
@@ -120,3 +124,90 @@ func (m *ErrorMessage) UnmarshalSSZ(buf []byte) error {
+}
+
+// UnmarshalSSZ unmarshals the provided bytes buffer into the
+// block by roots request object.
```
```suggestion
// blob sidecars by roots request object.
```

terencechain commented on beacon-chain/p2p/types/types.go:
```diff
@@ -120,3 +124,90 @@ func (m *ErrorMessage) UnmarshalSSZ(buf []byte) error {
+	return len(s)
+}
+
+func init() {
+	sizer := &eth.BlobIdentifier{}
+	blobIdSize = sizer.SizeSSZ()
+}
```
should we put init() at the start of the file?

kasey replied:
This is consistent with how I've use init elsewhere, always at the end of the file. It's not a strongly held opinion, but keeping it at the end would be consistent. My reasoning is that it's usually not the most interesting thing in the file, it only runs once to do basic initialization of values that tend to be declared at the top, and it is evaluated after everything else in the file (for instance, any other package scoped var declarations).

terencechain commented on beacon-chain/sync/blobs_test.go:
```diff
@@ -0,0 +1,340 @@
+		Timestamp:     0,
+		ExtraData:     make([]byte, 0),
+		BaseFeePerGas: bytesutil.PadTo([]byte("baseFeePerGas"), fieldparams.RootLength),
+		ExcessDataGas: bytesutil.PadTo([]byte("excessDataGas"), fieldparams.RootLength),
```
excess data gas is uint64 now and there's a new field data gas used

terencechain commented on beacon-chain/sync/blobs_test.go:
```diff
@@ -0,0 +1,340 @@
+	cleanup := func() {
+		require.NoError(t, undo())
+	}
+	maxBlobs := int(params.BeaconConfig().MaxBlobsPerBlock)
```
`MaxBlobsPerBlock` moved to field params

terencechain commented on cmd/beacon-chain/flags/base.go:
```diff
@@ -164,6 +164,18 @@ var (
 		Usage: "The factor by which block batch limit may increase on burst.",
 		Value: 2,
 	}
+	// BlockBatchLimit specifies the requested block batch size.
```
```suggestion
	// BlockBatchLimit specifies the requested blob sidecar batch size.
```

terencechain commented on cmd/beacon-chain/flags/base.go:
```diff
@@ -164,6 +164,18 @@ var (
+		Usage: "The amount of blobs the local peer is bounded to request and respond to in a batch.",
+		Value: 8,
+	}
+	// BlockBatchLimitBurstFactor specifies the factor by which block batch size may increase.
```
```suggestion
	// BlockBatchLimitBurstFactor specifies the factor by which blob sidecar batch size may increase.
```

terencechain commented on beacon-chain/sync/rpc_send_request.go:
```diff
@@ -129,3 +134,65 @@ func SendBeaconBlocksByRootRequest(
+		return nil, err
+	}
+	log.WithField("topic", topic).Debug("Sending blob sidecar request")
+	stream, err := p2pApi.Send(ctx, req, topic, pid)
```
Should we validate request `req` beforehand? Mainly length check to not exceed `MaxRequestBlobsSidecars`

kasey commented:
@terencechain fixup commit addressing your feedback:  https://github.com/prysmaticlabs/prysm/pull/12420/commits/58bd57a59b95e9338d04c86d4a906b16f3d7d1be

nisdas commented on beacon-chain/sync/rpc_blob_sidecars_by_root.go:
```diff
@@ -0,0 +1,125 @@
+		return errors.New("message is not type BlobSidecarsByRootReq")
+	}
+
+	blobIdents := *ref
```
we should have an independent function that validates the request( like how its done with block by range)

kasey replied:
done here https://github.com/prysmaticlabs/prysm/pull/12420/commits/73404ccc65cd793afc2f5e8cd7417a87a442b202

nisdas commented on beacon-chain/sync/rpc_blob_sidecars_by_root.go:
```diff
@@ -0,0 +1,125 @@
+		}
+		s.rateLimiter.add(stream, 1)
+		root, idx := bytesutil.ToBytes32(blobIdents[i].BlockRoot), blobIdents[i].Index
+		if root != buff.root {
```
maybe its better to explicitly just check that it is not a zeroed 32-byte root ?

kasey replied:
You mean we should check to make sure that the request doesn't contain a zero-value block root?

This code is checking to see if the block root for the sidecar id being currently processed is the same as the last step of the iteration. That way we can look up all the sidecars for the given root at once (because they are currently stored together in the db) and iterate through them - this is why the idents are sorted by root `sort.Sort(blobIdents)`.

If we do get a request for a zero root, we'll skip it as a side effect of the other invariants - the zero root will sort to the beginning of the list, so the initial value of `buff.scs` will be a nil list, which will have a length of zero, so we'll skip it here:
```
		if idx >= uint64(len(buff.scs)) {
			continue
		}
```
We could check for it and explicitly skip it. But if things did somehow change such that these invariants no longer prevented it from being skipped, then we would fail to look up the zero root from the db, and just return an error. So to me it feels like clutter to check for it.

prestonvanloon commented on beacon-chain/sync/rpc_blob_sidecars_by_root.go:
```diff
@@ -0,0 +1,132 @@
+			return err
+		}
+
+		// Throttle request processing to no more than batchSize/sec.
+		if i != 0 && i%batchSize == 0 && ticker != nil {
+			<-ticker.C
+		}
+		s.rateLimiter.add(stream, 1)
```
Should you increment the rate limiter before a potentially blocking event? i.e. if the chan is buffered/full then it is blocking _before_ the rate limiter was incremented which could be an issue.

kasey replied:
good catch, will fix.

kasey replied:
there's no buffered chan or otherwise non-deterministic block here fwiw, this is just a ticker.

nisdas reviewed (approved):
LGTM
